_WKHitTestResult should expose a frameInfo property

jeffmapple opened the pull request:
#### 9c20db5a05181f63b0989b73da5faf057f0c2445
<pre>
_WKHitTestResult should expose a frameInfo property
<a href="https://bugs.webkit.org/show_bug.cgi?id=263729">https://bugs.webkit.org/show_bug.cgi?id=263729</a>
<a href="https://rdar.apple.com/116254673">rdar://116254673</a>

Reviewed by Chris Dumez and Alex Christensen.

Add a frameInfo property to _WKHitTestResult and the plumbing to support it.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::frame const):
Added, this is the same frame returned by WKBundleHitTestResultGetFrame().

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult frameInfo]):
Added. Note that we need the page in order to create the WKFrameInfo.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
(WebKit::WebHitTestResultData::frameInfoDataFromHitTestResult):
Added. Creates a WebFrame from a LocalFrame and uses it to get the FrameInfoData.

* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
ContextMenuElementInfoMac now keeps track of the page.

* Source/WebKit/UIProcess/API/APIHitTestResult.cpp:
(API::HitTestResult::create):
API::HitTestResult now keeps track of the page.

* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::page):
(API::HitTestResult::frameInfo const):
(API::HitTestResult::HitTestResult):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageContextMenuClient):
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction _hitTestResult]):
We require a page to create an API::HitTestResult.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo hitTestResult]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
(WebKit::UIDelegate::UIClient::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _webHitTestResult]):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):&apos;
Added a new HitTestResultWhenClickingInSubframe test. Verifies that right-clicking in an iframe
correctly sets frameInfo in _WKHitTestResult.

Canonical link: <a href="https://commits.webkit.org/270308@main">https://commits.webkit.org/270308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec27ea8d8f56318500c8551beb68e44146083a15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25096 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3637 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27211 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23033 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25364 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5341 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1073 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23291 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2669 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27790 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2361 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22595 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28726 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22900 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22950 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26538 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/611 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3656 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6019 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2749 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2645 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -196,6 +200,19 @@ std::optional<WebKit::SharedMemory::Handle> WebHitTestResultData::getImageShared
+    if (!coreFrame)
+        return std::nullopt;
+
+    auto webFrame = WebFrame::fromCoreFrame(*coreFrame);
```
`RefPtr webFrame = ...`

jeffmapple replied:
I've changed all the places you noted to use explicit data types.

cdumez commented on Source/WebKit/Shared/WebHitTestResultData.cpp:
```diff
@@ -196,6 +200,19 @@ std::optional<WebKit::SharedMemory::Handle> WebHitTestResultData::getImageShared
 
+std::optional<FrameInfoData> WebHitTestResultData::frameInfoDataFromHitTestResult(const WebCore::HitTestResult& hitTestResult)
+{
+    auto* coreFrame = hitTestResult.frame();
```
`RefPtr coreFrame = ...`

cdumez commented on Source/WebKit/UIProcess/API/APIHitTestResult.h:
```diff
@@ -72,13 +73,19 @@ class HitTestResult : public API::ObjectImpl<API::Object::Type::HitTestResult> {
 
+    WebKit::WebPageProxy& page() { return m_page.get(); }
+
+    std::optional<WebKit::FrameInfoData> frameInfo() const { return m_data.frameInfo; }
```
Should return a `const std::optional<WebKit::FrameInfoData>&`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
```diff
@@ -246,8 +246,14 @@ - (_WKHitTestResult *)_hitTestResult
+    auto* sourceFrame = _navigationAction->sourceFrame();
+    if (!sourceFrame)
+        return nil;
+    auto* page = sourceFrame->page();
```
`RefPtr page = ...`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
```diff
@@ -246,8 +246,14 @@ - (_WKHitTestResult *)_hitTestResult
     auto& webHitTestResultData = _navigationAction->webHitTestResultData();
     if (!webHitTestResultData)
         return nil;
+    auto* sourceFrame = _navigationAction->sourceFrame();
```
`RefPtr sourceFrame = ...`

cdumez commented on Source/WebKit/Shared/WebHitTestResultData.h:
```diff
@@ -118,6 +120,8 @@ struct WebHitTestResultData {
 
     WebCore::IntRect elementBoundingBoxInWindowCoordinates(const WebCore::HitTestResult&);
 
+    std::optional<FrameInfoData> frameInfoDataFromHitTestResult(const WebCore::HitTestResult&);
```
Can this be marked as static? I don't think it uses any data members.

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -1805,7 +1805,7 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
             }
 
-            auto apiHitTestResult = API::HitTestResult::create(data);
+            auto apiHitTestResult = API::HitTestResult::create(data, page);
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -999,7 +999,7 @@ void WKPageSetPageContextMenuClient(WKPageRef pageRef, const WKPageContextMenuCl
             WKArrayRef newMenu = nullptr;
             if (m_client.base.version >= 2) {
-                auto webHitTestResult = API::HitTestResult::create(hitTestResultData);
+                auto webHitTestResult = API::HitTestResult::create(hitTestResultData, page);
```
`auto` -> `Ref`

cdumez commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -980,7 +980,7 @@ void WKPageSetPageContextMenuClient(WKPageRef pageRef, const WKPageContextMenuCl
             if (m_client.base.version >= 4 && m_client.getContextMenuFromProposedMenuAsync) {
                 auto proposedMenuItems = toAPIObjectVector(proposedMenuVector);
-                auto webHitTestResult = API::HitTestResult::create(hitTestResultData);
+                auto webHitTestResult = API::HitTestResult::create(hitTestResultData, page);
```
`auto` -> `Ref`

cdumez reviewed (approved):
LGTM with nits.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
```diff
@@ -30,6 +30,7 @@
 #include "APIObject.h"
 #include "ContextMenuContextData.h"
 #include "WebHitTestResultData.h"
+#include "WebPageProxy.h"
```
Making an explicit destructor in the cpp would make it so we only need a forward declaration here.

jeffmapple replied:
> Making an explicit destructor in the cpp would make it so we only need a forward declaration here.

It appears I can just remove the import from here and `APIHitTestResult.h` without an explicit destructor (although I did need to explicitly import `WebPageProxy.h` in `_WKHitTestResult.mm`).

jeffmapple replied:
I think I spoke too soon, my change to `APIHitTestResult.h` broke the iOS build.

jeffmapple committed:
_WKHitTestResult should expose a frameInfo property
https://bugs.webkit.org/show_bug.cgi?id=263729
rdar://116254673

Reviewed by Chris Dumez and Alex Christensen.

Add a frameInfo property to _WKHitTestResult and the plumbing to support it.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::frame const):
Added, this is the same frame returned by WKBundleHitTestResultGetFrame().

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult frameInfo]):
Added. Note that we need the page in order to create the WKFrameInfo.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
(WebKit::WebHitTestResultData::frameInfoDataFromHitTestResult):
Added. Creates a WebFrame from a LocalFrame and uses it to get the FrameInfoData.

* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/UIProcess/API/APIContextMenuElementInfoMac.h:
ContextMenuElementInfoMac now keeps track of the page.

* Source/WebKit/UIProcess/API/APIHitTestResult.cpp:
(API::HitTestResult::create):
API::HitTestResult now keeps track of the page.

* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::page):
(API::HitTestResult::frameInfo const):
(API::HitTestResult::HitTestResult):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageContextMenuClient):
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction _hitTestResult]):
We require a page to create an API::HitTestResult.

* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo hitTestResult]):
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::ContextMenuClient::menuFromProposedMenu):
(WebKit::UIDelegate::UIClient::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::mouseDidMoveOverElement):
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _webHitTestResult]):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):'
Added a new HitTestResultWhenClickingInSubframe test. Verifies that right-clicking in an iframe
correctly sets frameInfo in _WKHitTestResult.

Canonical link: https://commits.webkit.org/270308@main
